File: src/Systems/Swift/example2.swift
import Foundation

// Abstract Product
protocol Database {
    func connect()
    func query()
}

// Concrete Product
class Postgres: Database {
    func connect() {
        print("Connecting to PostgreSQL")
    }
    
    func query() {
        print("Querying PostgreSQL")
    }
}

class MySQL: Database {
    func connect() {
        print("Connecting to MySQL")
    }
    
    func query() {
        print("Querying MySQL")
    }
}

// Abstract Factory
protocol DatabaseFactory {
    func createDatabase() -> Database
}

// Concrete Factory
class PostgresFactory: DatabaseFactory {
    func createDatabase() -> Database {
        return Postgres()
    }
}

class MySQLFactory: DatabaseFactory {
    func createDatabase() -> Database {
        return MySQL()
    }
}

func useDatabase(factory: DatabaseFactory) {
    let db = factory.createDatabase()
    db.connect()
    db.query()
}

// Usage
useDatabase(factory: PostgresFactory())
useDatabase(factory: MySQLFactory())

File: src/Systems/Go/example2.go
package main

import "fmt"

// Abstract Product
type Database interface {
    Connect()
    Query()
}

// Concrete Product
type Postgres struct{}

func (p *Postgres) Connect() {
    fmt.Println("Connecting to PostgreSQL")
}

func (p *Postgres) Query() {
    fmt.Println("Querying PostgreSQL")
}

type MySQL struct{}

func (m *MySQL) Connect() {
    fmt.Println("Connecting to MySQL")
}

func (m *MySQL) Query() {
    fmt.Println("Querying MySQL")
}

// Abstract Factory
type DatabaseFactory interface {
    CreateDatabase() Database
}

// Concrete Factory
type PostgresFactory struct{}

func (pf *PostgresFactory) CreateDatabase() Database {
    return &Postgres{}
}

type MySQLFactory struct{}

func (mf *MySQLFactory) CreateDatabase() Database {
    return &MySQL{}
}

func useDatabase(factory DatabaseFactory) {
    db := factory.CreateDatabase()
    db.Connect()
    db.Query()
}

func main() {
    postgresFactory := &PostgresFactory{}
    mysqlFactory := &MySQLFactory{}
    useDatabase(postgresFactory)
    useDatabase(mysqlFactory)
}
